### PR TITLE
drivers: spi: convert spi_cs_control to use gpio_dt_spec

### PR DESCRIPTION
**Pull request by bbilas**
This MR contains changes and improvements related to gpio_dt_spec struct use within spi_cs_control struct.

It's based on the @JordanYates work that has been done here https://github.com/csiro-wsn/zephyr/commits/210801_spi_cs and because I didn't want to duplicate the work I've decided to take these changes and send to upstream as it's blocking the https://github.com/zephyrproject-rtos/zephyr/pull/37325 merge request.

@JordanYates If you have any objections to this MR please let me know.

### CODE REVIEW

File: include/devicetree/spi.h [hunk 1]
```diff
@@ -116,6 +116,42 @@ extern "C" {
...
+ *                        <&gpio2 20 GPIO_ACTIVE_LOW>;
+ *
+ *             a: spi-dev-a@0 {
+ *                     reg = <0>;
+ *             };
+ *
+ *             b: spi-dev-b@1 {
+ *                     reg = <1>;
+ *             };
+ *     };
```
Comment (gmarull) [inline include/devicetree/spi.h:140]: can be enclosed in `@code{.devicetree}` and `@endcode`

File: include/devicetree/spi.h [hunk 2]
```diff
@@ -116,6 +116,42 @@ extern "C" {
...
+ *
+ * Example usage:
+ *
+ *     DT_SPI_DEV_CS_GPIOS_DT_SPEC_GET(DT_NODELABEL(a)) \
```
Comment (gmarull) [inline include/devicetree/spi.h:144]: can be enclosed in `@code{.c}` and `@endcode`

File: include/devicetree/spi.h [hunk 3]
```diff
@@ -116,6 +116,42 @@ extern "C" {
...
+ *           // { DT_NODELABEL(gpio1), 10, GPIO_ACTIVE_LOW }
+ *     DT_SPI_DEV_CS_GPIOS_DT_SPEC_GET(DT_NODELABEL(b)) \
+ *           // { DT_NODELABEL(gpio2), 20, GPIO_ACTIVE_LOW }
+ *
+ * @param spi_dev a SPI device node identifier
+ * @return @p gpio_dt_spec struct corresponding with spi_dev's chip select
```
Comment (gmarull) [inline include/devicetree/spi.h:150]: `@return #gpio_dt_spec`

File: include/drivers/spi.h [hunk 4]
```diff
@@ -124,22 +124,24 @@ extern "C" {
...
- *    transmission and before releasing the CS line
- * @param gpio_dt_flags is the devicetree flags corresponding to how the CS
- *    line should be driven. GPIO_ACTIVE_LOW/GPIO_ACTIVE_HIGH should be
+ * @param gpio devicetree specification of CS GPIO.
+ *    The device pointer can be set to NULL to fully inhibit CS control if
+ *    necessary. The GPIO flags GPIO_ACTIVE_LOW/GPIO_ACTIVE_HIGH should be
  *    equivalent to SPI_CS_ACTIVE_HIGH/SPI_CS_ACTIVE_LOW options in struct
  *    spi_config.
+ * @param delay is a delay in microseconds to wait before starting the
+ *    transmission and before releasing the CS line
```
Comment (gmarull) [inline include/drivers/spi.h:133]: `@param` is not the proper way of documenting structs (was wrong), please move to
```c
struct x {
    /** Documentation for foo. */
    int foo;
};
```

File: drivers/spi/spi_context.h [hunk 5]
```diff
@@ -185,15 +185,13 @@ gpio_dt_flags_t spi_context_cs_active_level(struct spi_context *ctx)
 
 static inline void spi_context_cs_configure(struct spi_context *ctx)
```
Comment (gmarull) [inline drivers/spi/spi_context.h:186]: This function should return `gpio_pin_configure_dt` error code.

**Review by gmarull — CHANGES_REQUESTED** [review]
LGTM, some changes requested (I think it's worth resolving some existing issues now that this is being improved).

File: include/drivers/spi.h (hunk 4 above)
Comment (bbilas) [reply include/drivers/spi.h]: Fixed.

File: include/devicetree/spi.h (hunk 3 above)
Comment (bbilas) [reply include/devicetree/spi.h]: Fixed

File: include/devicetree/spi.h [hunk 6]
```diff
@@ -116,6 +116,42 @@ extern "C" {
...
+ *                     reg = <1>;
+ *             };
+ *     };
+ *
+ * Example usage:
+ *
+ *     DT_SPI_DEV_CS_GPIOS_DT_SPEC_GET(DT_NODELABEL(a)) \
+ *           // { DT_NODELABEL(gpio1), 10, GPIO_ACTIVE_LOW }
+ *     DT_SPI_DEV_CS_GPIOS_DT_SPEC_GET(DT_NODELABEL(b)) \
+ *           // { DT_NODELABEL(gpio2), 20, GPIO_ACTIVE_LOW }
```
[the remark this answers is not in the document]
Comment (bbilas) [reply include/devicetree/spi.h]: Fixed

File: include/devicetree/spi.h [hunk 7]
```diff
@@ -116,6 +116,42 @@ extern "C" {
...
+ *
+ *             b: spi-dev-b@1 {
+ *                     reg = <1>;
+ *             };
+ *     };
+ *
+ * Example usage:
+ *
+ *     DT_SPI_DEV_CS_GPIOS_DT_SPEC_GET(DT_NODELABEL(a)) \
+ *           // { DT_NODELABEL(gpio1), 10, GPIO_ACTIVE_LOW }
```
[the remark this answers is not in the document]
Comment (bbilas) [reply include/devicetree/spi.h]: Fixed

File: include/devicetree/spi.h (hunk 2 above)
Comment (bbilas) [reply include/devicetree/spi.h:147]: Fixed

File: include/devicetree/spi.h (hunk 1 above)
Comment (bbilas) [reply include/devicetree/spi.h:141]: Fixed.

File: drivers/spi/spi_context.h (hunk 5 above)
Comment (bbilas) [reply drivers/spi/spi_context.h]: Fixed.

File: include/devicetree/spi.h [hunk 8]
```diff
@@ -116,6 +116,46 @@ extern "C" {
...
+ *     };
+ * @endcode
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *     DT_SPI_DEV_CS_GPIOS_DT_SPEC_GET(DT_NODELABEL(a)) \
+ *           // { DT_NODELABEL(a), 10, GPIO_ACTIVE_LOW }
+ *     DT_SPI_DEV_CS_GPIOS_DT_SPEC_GET(DT_NODELABEL(b)) \
+ *           // { DT_NODELABEL(b), 20, GPIO_ACTIVE_LOW }
```
Comment (gmarull) [inline include/devicetree/spi.h:150]: nit: `DEVICE_DT_GET(DT_NODELABEL(a/b))`

Comment (bbilas) [reply include/devicetree/spi.h:150]: Do you mean
<pre>
// { DEVICE_DT_GET(DT_NODELABEL(b)), 20, GPIO_ACTIVE_LOW }
</pre>
and
<pre>
// { DEVICE_DT_GET(DT_NODELABEL(b)), 20, GPIO_ACTIVE_LOW }
</pre> ?

Comment (gmarull) [reply include/devicetree/spi.h:150]: correct

Comment (bbilas) [reply include/devicetree/spi.h]: Done :)

File: include/drivers/spi.h [hunk 9]
```diff
@@ -124,22 +124,28 @@ extern "C" {
...
+	 */
+	union {
+		struct gpio_dt_spec gpio;
+		struct {
+			const struct device *gpio_dev;
+			gpio_pin_t gpio_pin;
+			gpio_dt_flags_t gpio_dt_flags;
+		};
+	};
+	/*
```
Comment (gmarull) [inline include/drivers/spi.h:144]:
```suggestion
	/**
```

File: include/drivers/spi.h [hunk 10]
```diff
@@ -124,22 +124,28 @@ extern "C" {
...
- *    line should be driven. GPIO_ACTIVE_LOW/GPIO_ACTIVE_HIGH should be
- *    equivalent to SPI_CS_ACTIVE_HIGH/SPI_CS_ACTIVE_LOW options in struct
- *    spi_config.
  */
 struct spi_cs_control {
-	const struct device	*gpio_dev;
-	uint32_t		delay;
-	gpio_pin_t		gpio_pin;
-	gpio_dt_flags_t		gpio_dt_flags;
+	/*
```
Comment (gmarull) [inline include/drivers/spi.h:129]:
```suggestion
	/**
```

**Review by gmarull — APPROVED** [review]

File: include/drivers/spi.h (hunk 9 above)
Comment (bbilas) [reply include/drivers/spi.h]: Done.

File: include/drivers/spi.h (hunk 10 above)
Comment (bbilas) [reply include/drivers/spi.h]: Done.

File: include/devicetree/spi.h [hunk 11]
```diff
@@ -116,6 +116,46 @@ extern "C" {
...
+ *     };
+ * @endcode
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *     DT_SPI_DEV_CS_GPIOS_DT_SPEC_GET(DT_NODELABEL(a)) \
+ *           // { DEVICE_DT_GET(DT_NODELABEL(a)), 10, GPIO_ACTIVE_LOW }
+ *     DT_SPI_DEV_CS_GPIOS_DT_SPEC_GET(DT_NODELABEL(b)) \
+ *           // { DEVICE_DT_GET(DT_NODELABEL(b)), 20, GPIO_ACTIVE_LOW }
```
Comment (MaureenHelm) [inline include/devicetree/spi.h:150]: Should this be `DT_NODELABEL(gpio2)`?

Comment (gmarull) [reply include/devicetree/spi.h:150]: Good catch, @MaureenHelm

Comment (bbilas) [reply include/devicetree/spi.h]: Fixed.

**Review by MaureenHelm — APPROVED** [review]

**Comment by JordanYates** [comment]
The following commit was not pulled in from the linked branch: 2cf39d97d7d33de4b0c73c6b6db7ca39c1c425cd.
Obviously it is not complete, but the fields need to be marked deprecated so that the old fields can be removed in the future.

**Comment by bbilas** [comment]
> The following commit was not pulled in from the linked branch: [2cf39d9](https://github.com/zephyrproject-rtos/zephyr/commit/2cf39d97d7d33de4b0c73c6b6db7ca39c1c425cd). Obviously it is not complete, but the fields need to be marked deprecated so that the old fields can be removed in the future.

That was intentional because some of the drivers are not converted to use the new gpio_dt_spec so the builds were failing due to _deprecated macro thus I've removed this commit from there. It could be done later when everything will be prepared to mark these fields deprecated.

**Comment by JordanYates** [comment]
> That was intentional because some of the drivers are not converted to use the new gpio_dt_spec so the builds were failing due to _deprecated macro thus I've removed this commit from there.

Deprecating the old fields was an important part of those commits, it should not be lost/forgotten about.

**Comment by bbilas** [comment]
> > That was intentional because some of the drivers are not converted to use the new gpio_dt_spec so the builds were failing due to _deprecated macro thus I've removed this commit from there.
> 
> Deprecating the old fields was an important part of those commits, it should not be lost/forgotten about.

Agreed, when https://github.com/zephyrproject-rtos/zephyr/pull/37498 will be merged we can push this commit to mark those fields deprecated.

**Comment by mbolivar-nordic** [comment]
@cfriedt I wish that this hadn't been merged without a devicetree maintainer ACKing the changes to include/devicetree/spi.h. That's not SPI API, that's DT API, and this new macro doesn't belong there.

Details in https://github.com/zephyrproject-rtos/zephyr/issues/42149. No harm done since this is unreleased; just making a note here.

**Pull request merged**